src/Workspaces/Remote/Core/SolutionAssetProvider.cs

```diff
@@ -52,58 +52,143 @@ public async ValueTask GetAssetsAsync(PipeWriter pipeWriter, Checksum solutionCh
                 assetMap = await scope.GetAssetsAsync(checksums, cancellationToken).ConfigureAwait(false);
             }
 
-            // We can cancel early, but once the pipe operations are scheduled we rely on both operations running to
-            // avoid deadlocks (the exception handler in 'task1' ensures progress is made in 'task2').
             cancellationToken.ThrowIfCancellationRequested();
-            var mustNotCancelToken = CancellationToken.None;
-
-            // Work around the lack of async stream writing in ObjectWriter, which is required when writing to the RPC
-            // pipe. Run two tasks - the first synchronously writes to a local pipe and the second asynchronously
-            // transfers the data to the RPC pipe.
-            //
-            // Configure the pipe to never block on write (waiting for the reader to read). This prevents deadlocks but
-            // might result in more (non-contiguous) memory allocated for the underlying buffers. The amount of memory
-            // is bounded by the total size of the serialized assets.
-            var localPipe = new Pipe(s_pipeOptionsWithUnlimitedWriterBuffer);
-
-            var task1 = Task.Run(() =>
+
+            using var stream = new PipeWriterStream(pipeWriter);
+            await RemoteHostAssetSerialization.WriteDataAsync(
+                stream, singleAsset, assetMap, serializer, scope.ReplicationContext,
+                solutionChecksum, checksums, cancellationToken).ConfigureAwait(false);
+
+            // Ensure any last data written into the stream makes it into the pipe.
+            await stream.FlushAsync(cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Simple port of
+        /// https://github.com/AArnott/Nerdbank.Streams/blob/dafeb5846702bc29e261c9ddf60f42feae01654c/src/Nerdbank.Streams/BufferWriterStream.cs#L16.
+        /// Wraps a <see cref="PipeWriter"/> in a <see cref="Stream"/> interface.  Preferred over <see
+        /// cref="PipeWriter.AsStream(bool)"/> as that API produces a stream that will synchronously flush after
+        /// <em>every</em> write.  That's undesirable as that will then block a thread pool thread on the actual
+        /// asynchronous flush call to the underlying PipeWriter
+        /// </summary>
+        private class PipeWriterStream : Stream, IDisposableObservable
```

Comment:
it would be nice if this type already existed with the semantics listed above.  However, it's really barely any code for us to own here, so i don't mind the custom stream implementation.  All the reading methods are no-ops, and the writing methods are trivial pass-throughs to the efficient writing methods on the pipe.